Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift

```diff
@@ -0,0 +1,467 @@
+        case forwardResponseEnd(readPending: Bool, clearReadTimeoutTimer: Bool)
+
+        case failRequest(Error, closeStream: Bool)
+
+        case read
+        case wait
+    }
+
+    private var state: State = .initialized
+
+    private var isChannelWritable: Bool
+    private let idleReadTimeout: TimeAmount?
+
+    init(isChannelWritable: Bool, idleReadTimeout: TimeAmount?) {
+        self.isChannelWritable = isChannelWritable
+        self.idleReadTimeout = idleReadTimeout
+    }
+
+    mutating func writabilityChanged(writable: Bool) -> Action {
+        self.isChannelWritable = writable
+
+        switch self.state {
+        case .initialized,
+             .finished,
+             .failed:
+            return .wait
+
+        case .running(.verifyRequest, _), .running(.endSent, _):
+            return .wait
+
+        case .running(.streaming(let expectedBody, let sentBodyBytes, producer: .paused), let responseState):
+            if writable {
+                let requestState: RequestState = .streaming(
+                    expectedBodyLength: expectedBody,
+                    sentBodyBytes: sentBodyBytes,
+                    producer: .producing
+                )
+
+                self.state = .running(requestState, responseState)
+                return .resumeRequestBodyStream
+            } else {
+                // no state change needed
+                return .wait
+            }
+
+        case .running(.streaming(let expectedBodyLength, let sentBodyBytes, producer: .producing), let responseState):
+            if !writable {
+                let requestState: RequestState = .streaming(
+                    expectedBodyLength: expectedBodyLength,
+                    sentBodyBytes: sentBodyBytes,
+                    producer: .paused
+                )
+                self.state = .running(requestState, responseState)
+                return .pauseRequestBodyStream
+            } else {
+                // no state change needed
+                return .wait
+            }
+        }
+    }
+
+    mutating func readEventCaught() -> Action {
+        return .read
+    }
+
+    mutating func errorHappened(_ error: Error) -> Action {
+        switch self.state {
+        case .initialized:
+            preconditionFailure("After the state machine has been initialized, start must be called immediately. Thus this state is unreachable")
+        case .running:
+            self.state = .failed(error)
+            return .failRequest(error, closeStream: true)
+        case .finished, .failed:
+            preconditionFailure("If the request is finished or failed, we expect the connection state machine to remove the request immediately from its state. Thus this state is unreachable.")
+        }
+    }
+
+    mutating func start() -> Action {
+        guard case .initialized = self.state else {
+            preconditionFailure("Invalid state")
```

Comment:
A comment indicating that `start()` must be called first and exactly once would be useful.